aoc2022/src/main/scala/day17.scala

```diff
@@ -0,0 +1,110 @@
+    
+    case Min extends Rock((x: Int, y: Int) => Vector(Point(x, y), Point(x+1, y), Point(x+2, y), Point(x+3, y)), 4, 1)
+    case Plus extends Rock((x: Int, y: Int) => Vector(Point(x, y+1), Point(x+1, y), Point(x+1, y+1), Point(x+1, y+2), Point(x+2, y+1)), 3, 3)
+    case L extends Rock((x: Int, y: Int) => Vector(Point(x, y), Point(x+1, y), Point(x+2, y), Point(x+2, y+1), Point(x+2, y+2)), 3, 3)
+    case Pipe extends Rock((x: Int, y: Int) => Vector(Point(x, y), Point(x, y+1), Point(x, y+2), Point(x, y+3)), 1, 4)
+    case Block extends Rock((x: Int, y: Int) => Vector(Point(x, y), Point(x, y+1), Point(x+1, y), Point(x+1, y+1)), 2, 2)
+
+  object Rock:
+    val rocks: Vector[Rock] = Vector(Min, Plus, L, Pipe, Block)
+    val nRocks: Int = rocks.length
+    val jets: Vector[Char] = input
+    val nJets: Int = jets.length
+    
+    def simulate(yMax: Int, ij: (Int, Int), obstacles: List[Point], n: Int): Int =
+      
+      def go(r: Rock, x: Int, y: Int, j: Int): (Int, Int, Int) =
+        val xs: Int = r.blow(jets(j), x, y, obstacles)
+        val ys: Int = r.down(xs, y, obstacles)
+        if y == ys then (xs, y, j) else go(r, xs, ys, (j + 1) % nJets)
+      
+      // enable println below to see repetition:
+      //if ij._2 == 3 then println(s"$yMax and $n")
+      
+      if n <= 0 then yMax
+      else
+        val thisRock: Rock = rocks(ij._1)
+        val (xf, yf, jf): (Int, Int, Int) = go(thisRock, 2, yMax+4, ij._2)
+        val settledRock: Vector[Point] = thisRock.loc(xf, yf)
+        val nextMaxY: Int = if (yf + thisRock.height - 1) > yMax then yf + thisRock.height - 1 else yMax
+        
+        // Limit the obstacles List to max 100 rocks, because the rocks below that don't influence next rocks. This 
+        // increases program speed and memory efficiency. 
+        val nextObs: List[Point] = settledRock.foldLeft(obstacles)((b, a) => a :: b).take(100)  
+        simulate(nextMaxY, ((ij._1 + 1) % nRocks, (jf + 1) % nJets), nextObs, n - 1)
+  
+  private val answer1: Int = Rock.simulate(0, (0, 0), List.empty[Point], 2022)
+  println(s"Answer day $day part 1: ${answer1} [${System.currentTimeMillis - start1}ms]")
+  
+  private val start2: Long =
+    System.currentTimeMillis
+
+  /**
+   * 
+   * After 1 iteration through the jets, the next iterations start a loop with an increase in height of 2768 each 
+   * 1755 blocks. This can be found by seeing that the block index i and the jet index j repeat at i=0 and j=3 
+   * (in my case). Thus, the start n (before this repeated loop starts) is 1752 with an height of 2729. 
+   * 
+   * The remainder after running the repeated loop is (1000000000000L - 1752) % 1755 = 1408. 
+   * The height of 1408 blocks in the repeated loop is 2233 (can be computed with function above: 
+   * score = sim(n=1752 + 1408) - 2729). 
+   * 
+   * Thus the final score can be computed with the formula below:
+   */
+
+  private val answer2: Long = (1000000000000L - 1752) / 1755 * 2768 + 2729 + 2233
```

Comment:
you might like [this](https://github.com/nmcb/aoc2022/blob/main/src/main/scala/Day17.scala#L145-L186) utility :)

Reply:
Yep, I like that utility 👍 

Copied and included in my utills package